### PR TITLE
Add profiles to manage docker-compose containers

### PR DESCRIPTION
**Pull request by Tristan-WorkGH**
Long awaited [profiles](https://docs.docker.com/compose/profiles/) to manage selectively containers!

Some notes:
  * I have keep `docker-compose.yml` splited.
  * The list of profiles is the same than with directories (`study`,`suite`,`merging`,`dynmap`).
    kibana and pgadmin are isolated in their own profile as they are rarely used by devs.
    An `all` profile is also added to be able to do a `docker compose --profile all down`.
  * The list of profiles with what they include is also in the readme.
  * These feature of docker-compose have been used:
    * [profiles](https://docs.docker.com/compose/profiles/) to be able to manage groups of containers from CLI

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -192,6 +195,30 @@ http://localhost:5601
 In order to show documents in the case-server index with Kibana, you must first create the index pattern ('Management' page) : case-server*
 
 
+### Installing / Updating docker-compose to v2
+Docker-compose v2 is necessary to have to profiles feature.  
+If possible, prefer to install it with your package manager if your on a Unix system.
```
Comment (jonenst) [inline README.md:200]: you're or you are

File: README.md [hunk 2]
```diff
@@ -192,6 +260,30 @@ http://localhost:5601
 In order to show documents in the case-server index with Kibana, you must first create the index pattern ('Management' page) : case-server*
 
 
+### Installing / Updating docker-compose to v2
+Docker-compose v2 is necessary to have to profiles feature.
```
Comment (flomillot) [inline README.md:264]:
```suggestion
Docker compose v2 is necessary to use the profiles feature.  
```

File: README.md [hunk 3]
```diff
@@ -192,6 +260,30 @@ http://localhost:5601
...
 
+### Installing / Updating docker-compose to v2
+Docker-compose v2 is necessary to have to profiles feature.  
+If possible, prefer to install it with your package manager if you are on a Unix system.
+
+> [!NOTE]  
+> You need a client (docker-cli) of ~v19~ v20 at least to have the system of cli-plugins.
+> It isn't necessary to update your Docker engine or client else.
+
+You can install docker-compose with these commands, as instructed in [the doc](https://docs.docker.com/compose/install/linux/#install-the-plugin-manually) and the [migration guide](https://docs.docker.com/compose/migrate/):
```
Comment (flomillot) [inline README.md:271]:
```suggestion
You can install docker compose with these commands, as instructed in [the doc](https://docs.docker.com/compose/install/linux/#install-the-plugin-manually) and the [migration guide](https://docs.docker.com/compose/migrate/):
```

File: README.md [hunk 4]
```diff
@@ -192,6 +260,30 @@ http://localhost:5601
...
+```shell
+curl -LR --create-dirs -o $HOME/.local/bin/docker-compose https://github.com/docker/compose/releases/latest/download/docker-compose-linux-x86_64
+#wget -x -O $HOME/.local/bin/docker-compose https://github.com/docker/compose/releases/latest/download/docker-compose-linux-x86_64
+chmod +x $HOME/.local/bin/docker-compose
+mkdir -p $HOME/.docker/cli-plugins
+ln -s $HOME/.local/bin/docker-compose $HOME/.docker/cli-plugins/docker-compose
+docker compose version
+```
+You must get this output from docker compose now:
+> Docker Compose version v2.20.3
```
Comment (flomillot) [inline README.md:281]:
```suggestion
> Docker Compose version v2.X.X
```

File: README.md [hunk 5]
```diff
@@ -31,43 +31,111 @@ To do this, you must copy the following files in the init directory :
...
+```shell
+$ cd docker-compose
+$ docker compose --profile suite <cmd>
+```
+
+You can also combine multiple profiles:
+```shell
+$ docker compose --profile study --profile mapping <cmd>
+```
+But please note that services/container who belongs to at least one profile can't be accessed if the profile isn't specified.
```
Comment (flomillot) [inline README.md:115]: Why do you say this ? For me it works. 
It's just that you need to be careful for the dependencies according to the doc (`depends_on`).
I missed something ?

**Review by flomillot — COMMENTED** [review]
I'm not sure we should have an `all` profile.
According to the doc :
_"The core services of your application shouldn't be assigned profiles so they are always enabled and automatically started."_
Otherwise, everything seems to work for me.

edit : nevermind, `all` profile is just to be able to down everything easily as you say in the description, actually services in base does not have profiles

File: README.md (hunk 5 above)
Comment (flomillot) [reply README.md:115]: Actually `docker compose up` start everything but `down` does not work well if you don't specify the `all` profile...
But `docker compose stop` still works well.

File: README.md [hunk 6]
```diff
@@ -31,43 +31,111 @@ To do this, you must copy the following files in the init directory :
...
+
+To use a profile, you use simply:
+```shell
+$ cd docker-compose
+$ docker compose --profile suite <cmd>
+```
+
+You can also combine multiple profiles:
+```shell
+$ docker compose --profile study --profile mapping <cmd>
```
Comment (jonenst) [inline README.md:113]: mapping -> dynmap ?

File: docker-compose/.env [hunk 7]
```diff
@@ -3,4 +3,8 @@ PROJECT_MERGING_DIR_NAME=merging
 PROJECT_DYNAMIC_MAPPING_DIR_NAME=dynamic-mapping
 PROJECT_SUITE_DIR_NAME=suite
 PROJECT_TECHNICAL_DIR_NAME=technical
-DATABASE_PREFIX_NAME=
+#DATABASE_PREFIX_NAME=
+#PROJECT_DIR_NAME=merging/study/suite --> will block postgres container if geo-data-server or cgmes-boundary-server are unavailable
```
[the remark this answers is not in the document]
Comment (Tristan-WorkGH) [reply docker-compose/.env:7]: yes:
* `DATABASE_PREFIX_NAME` is define in the user/host environment (it's in the doc)
* `PROJECT_DIR_NAME` is define line 8, the line 7 is a note. I have found a way in another branch, will be another PR.

File: README.md (hunk 4 above)
Comment (Tristan-WorkGH) [reply README.md:281]: That is just an example of the output, i don't think it's necessary... no?

File: README.md (hunk 3 above)
Comment (Tristan-WorkGH) [reply README.md:271]: It's the name of [the tool/file](https://github.com/docker/compose).

**Comment by flomillot** [comment]
> > @flomillot
> > [...]
> > edit 2 : we lose the global log output in lazydocker :(
> 
> I've never use this tool, but I think it come from the update of docker-compose because the logs are from docker-compose who merge each container logs from docker. And I change nothing about the output of the services.

I use docker compose V2 since the start, but not a big deal anyway, maybe I can fix it

File: README.md (hunk 4 above)
Comment (flomillot) [reply README.md:281]: For me it's better, otherwise we can think that we need this version specifically

**Comment by Tristan-WorkGH** [comment]
> > > @flomillot
> > > [...]
> > > edit 2 : we lose the global log output in lazydocker :(
> > 
> > 
> > I've never use this tool, but I think it come from the update of docker-compose because the logs are from docker-compose who merge each container logs from docker. And I change nothing about the output of the services.
> 
> I use docker compose V2 since the start, but not a big deal anyway, maybe I can fix it

Try to update lazydocker to last version, I just try and I have the logs.
image

File: README.md (hunk 3 above)
Comment (flomillot) [reply README.md:271]: Well the fille yes, sure, but the tool is called docker compose now, docker-compose is the previous version. We avoid confusion then.

File: README.md (hunk 5 above)
Comment (Tristan-WorkGH) [reply README.md:115]: Hum, maybe I wasn't clear in the doc.  
What I want to say is if you do for example `docker compose stop study-server` or `docker compose --profile merging stop study-server` it would not work because study-server is in the `study` and `suite` profiles, so not visible until one of the profiles it is attached (`study` or `suite` ... or `all`) is passed in the cli.

**Comment by flomillot** [comment]
> > > > @flomillot
> > > > [...]
> > > > edit 2 : we lose the global log output in lazydocker :(
> > > 
> > > 
> > > I've never use this tool, but I think it come from the update of docker-compose because the logs are from docker-compose who merge each container logs from docker. And I change nothing about the output of the services.
> > 
> > 
> > I use docker compose V2 since the start, but not a big deal anyway, maybe I can fix it
> 
> Try to update lazydocker to last version, I just tried and I have the logs. image

If you click on "docker-compose" under Projet, you don't have the logs. With the previous version, you can see the logs of all the services.
But forget it, it's not a big deal, I don't think we are a lot using this tool anyway.

File: README.md (hunk 5 above)
Comment (flomillot) [reply README.md:115]: I tried the first one and it works for me.

Comment (Tristan-WorkGH) [reply README.md:115]: After some tests, the [auto-starting-profiles](https://docs.docker.com/compose/profiles/#auto-starting-profiles-and-dependency-resolution) documentation seems to be outdated
  * `up` and `down` **need** profiles explicitly
  * `start`/`stop`/`restart` _don't care_ of profiles, they only need the container(s) to be created by a previously `up`
    * and check `depends_on`

➡️ Need to update the readme

File: docker-compose/docker-compose.base.yml [hunk 8]
```diff
@@ -1,56 +1,51 @@
...
     image: powsybl/case-server:latest
     ports:
       - 5000:80
     volumes:
-      - $PWD/../../k8s/resources/common/config/case-server-application.yml:/config/specific/application.yml:Z
-      - $PWD/../../k8s/resources/common/config/common-application.yml:/config/common/application.yml:Z
-      - $GRIDSUITE_DATABASES/cases:/cases:Z
+      - ./../k8s/resources/common/config/case-server-application.yml:/config/specific/application.yml:ro
+      - ./../k8s/resources/common/config/common-application.yml:/config/common/application.yml:ro
+      - ${GRIDSUITE_DATABASES:-/tmp/gridsuite}/cases:/cases
```
Comment (flomillot) [inline docker-compose/docker-compose.base.yml:10]: Why did you have to change this line and add `-/tmp/gridsuite` ?

File: README.md [hunk 9]
```diff
@@ -31,43 +31,115 @@ To do this, you must copy the following files in the init directory :
...
+```shell
+$ docker compose --profile study --profile dynmap <cmd>
+```
+
+> [!IMPORTANT]  
+> Contrary to what the Docker documentation say:
+>   * With the commands `up` & `down`, services/container who belongs to at least one profile can't be accessed if the profile isn't specified.
+>     For example `docker compose up study-server` would not work because the profile `study` isn't passed in the CLI.
+>     The correct CLI would be `docker compose --profile study up study-server`.
+>   * With the commands `start`, `stop`, `restart`, the `--profile ...` as no effect because theses commands affect the containers already created by a previous `up` command.
```
Comment (jonenst) [inline README.md:121]:
```suggestion
>   * With the commands `start`, `stop`, `restart`, the `--profile ...` has no effect because theses commands affect the containers already created by a previous `up` command.
```

File: docker-compose/.env (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tristan-WorkGH) [reply docker-compose/.env:7]: I have fixed for `DATABASE_PREFIX_NAME` ; and for `PROJECT_DIR_NAME`, I have another PR to fix the problem.

File: docker-compose/docker-compose.base.yml (hunk 8 above)
Comment (Tristan-WorkGH) [reply docker-compose/docker-compose.base.yml]: will be in another PR as asked

File: docker-compose/docker-compose.base.yml [hunk 10]
```diff
@@ -1,85 +1,118 @@
-version: '3'
-
+version: '3.9'
```
Comment (jonenst) [inline docker-compose/docker-compose.base.yml:1]: why 3.9 ? it doesn't exist
https://docs.docker.com/compose/compose-file/04-version-and-name/#version-top-level-element
https://docs.docker.com/compose/compose-file/compose-versioning/#compatibility-matrix

File: docker-compose/docker-compose.base.yml [hunk 11]
```diff
@@ -1,85 +1,118 @@
...
+      - study
+      - study-light
+      - merging
+      - dynmap
+      - import
     image: powsybl/case-server:latest
     ports:
       - 5000:80
     volumes:
-      - $PWD/../../k8s/resources/common/config/case-server-application.yml:/config/specific/application.yml:Z
```
Comment (jonenst) [inline docker-compose/docker-compose.base.yml:10]: why this change ?

File: docker-compose/docker-compose.base.yml (hunk 10 above)
Comment (Tristan-WorkGH) [reply docker-compose/docker-compose.base.yml:1]: true 😮
It seem that during a long time [there was confusion](https://forums.docker.com/t/docker-compose-version-3-8-or-3-9-for-latest/102439) ([with inconsistencies in the doc](https://github.com/docker/docs/issues/12214))  
but in the end, [3.9 never existed](https://github.com/docker/docs/pull/17390)

File: README.md [hunk 12]
```diff
@@ -47,60 +47,18 @@ Install the orchestration tool [Docker Compose v2](https://github.com/docker/com
...
+| gridmerge&#8209;app<br/>balances&#8209;adjustment&#8209;server<br/>case&#8209;import&#8209;job<br/>case&#8209;validation&#8209;server<br/>cgmes&#8209;assembling&#8209;job<br/>cgmes&#8209;boundary&#8209;import&#8209;job<br/>cgmes&#8209;boundary&#8209;server<br/>merge&#8209;notification&#8209;server<br/>merge&#8209;orchestrator&#8209;server | | ✅ | | | | | ✅ | | | |
+| gridstudy&#8209;app<br/>dynamic&#8209;simulation&#8209;server<br/>timeseries&#8209;server | | | ✅ | | | ✅ | ✅ | | | |
+| cgmes&#8209;gl&#8209;server<br/>odre&#8209;server<br/>security&#8209;analysis&#8209;server<br/>sensitivity&#8209;analysis&#8209;server<br/>shortcircuit&#8209;server<br/>voltage&#8209;init&#8209;server | | | ✅ | | | | ✅ | | | |
+| directory&#8209;notification&#8209;server<br/>directory&#8209;server<br/>explore&#8209;server<br/>geo&#8209;data&#8209;server<br/>gridexplore&#8209;app<br/>network&#8209;map&#8209;server<br/>network&#8209;modification&#8209;server<br/>single&#8209;line&#8209;diagram&#8209;server<br/>study&#8209;notification&#8209;server<br/>study&#8209;server | | | ✅ | ✅ | | ✅ | ✅ | | | |
+| case&#8209;import&#8209;server | | | | | | | | ✅ | | |
```
Comment (flomillot) [inline README.md:61]: IntelliJ propose to reformat this to make it more readable for maintainers :
<details><summary>Details</summary>
<p>

image
image

</p>
</details>

**Review by flomillot — COMMENTED** [review]
Warning when I launch or stop the stack, everytime : 
`WARN[0000] The "DATABASE_PREFIX_NAME" variable is not set. Defaulting to a blank string.`
Maybe use an empty string as default to avoid this message ? Or something else I don't know

Btw I think there is a problem with the databases, when you start the stack there is some error messages like they can't create the dabatases because they already exists then the app don't works.


<details><summary>Details</summary>
<p>


image

</p>
</details>

File: README.md (hunk 12 above)
Comment (Tristan-WorkGH) [reply README.md:61]: yes, I did this because the lines is too long to be readable in the "more readable" form with word warping...

**Comment by flomillot** [comment]
> > Warning when I launch or stop the stack, everytime : `WARN[0000] The "DATABASE_PREFIX_NAME" variable is not set. Defaulting to a blank string.` Maybe use an empty string as default to avoid this message ? Or something else I don't know
> 
> Yeah, that because it isn't define in your environment variables (it's an optional variable). In docker-compose v1, this warning didn't exist, it was silent in this case. 

I don't have this warning on the main branch with DC V2.
We have it because you modified this line no ?
`DATABASE_PREFIX_NAME=$DATABASE_PREFIX_NAME`
Can't you use a blank string as default value ? It's what docker use anyway if the var is missing.

**Review by flomillot — APPROVED** [review]

File: docker-compose/docker-compose.base.yml (hunk 11 above)
Comment (Tristan-WorkGH) [reply docker-compose/docker-compose.base.yml:10]: for another pr...

**Review by EstherDarkish — APPROVED** [review]

**Review by jonenst — APPROVED** [review]
